datalad/distribution/siblings.py

```diff
@@ -0,0 +1,382 @@
+        publish_depends=publish_depends,
+        publish_by_default=publish_by_default,
+        annex_wanted=annex_wanted_opt,
+        annex_group=annex_group_opt,
+        annex_groupwanted=annex_groupwanted_opt,
+        inherit=inherit_opt,
+
+        recursive=recursion_flag,
+        recursion_limit=recursion_limit)
+
+    @staticmethod
+    @datasetmethod(name='siblings')
+    @eval_results
+    def __call__(
+            mode='query',
+            dataset=None,
+            name=None,
+            url=None,
+            pushurl=None,
+            # TODO consider true, for now like add_sibling
+            fetch=False,
+            as_common_datasrc=None,
+            publish_depends=None,
+            publish_by_default=None,
+            annex_wanted=None,
+            annex_group=None,
+            annex_groupwanted=None,
+            inherit=False,
+            recursive=False,
+            recursion_limit=None):
+
+        # TODO catch invalid mode specified
+        mode_worker_map = {
+            'query': _query_remotes,
+            'add': _add_remote,
+            'configure': _configure_remote,
+            'remove': _remove_remote,
+        }
+        # all worker strictly operate on a single dataset
+        # anything that deals with hierarchies and/or dataset
+        # relationships in general should be dealt with in here
+        # at the top-level and vice versa
+        worker = mode_worker_map[mode]
+
+        dataset = require_dataset(
+            dataset, check_installed=False, purpose='sibling configuration')
+        refds_path = dataset.path
+
+        res_kwargs = dict(refds=refds_path)
+
+        ds_name = basename(dataset.path)
+
+        # do not form single list of datasets (with recursion results) to
+        # give fastest possible response, for the precise of a long-all
+        # function call
+        ds = dataset
+        for r in worker(
+                # always copy signature to below to avoid bugs!
+                ds, name,
+                # for top-level dataset there is no layout questions
+                _mangle_urls(url, ds_name),
+                _mangle_urls(pushurl, ds_name),
+                fetch,
+                as_common_datasrc, publish_depends, publish_by_default,
+                annex_wanted, annex_group, annex_groupwanted,
+                inherit,
+                **res_kwargs):
+            yield r
+        if not recursive:
+            return
+
+        # do we have instructions to register siblings with some alternative
+        # layout?
+        replicate_local_structure = url and "%NAME" not in url
+
+        for subds in dataset.subdatasets(
+                fulfilled=True,
+                recursive=recursive, recursion_limit=recursion_limit,
+                result_xfm='datasets'):
+            subds_name = relpath(subds.path, start=dataset.path)
+            if replicate_local_structure:
+                subds_url = _urljoin(url, subds_name)
```

Comment:
Related to the comment about confusing urljoins: We need to consider non-POSIX paths mixed into urls.
At least, `_urljoin` has to use `os.separator` instead of `'/'`.